fluent-kafka-streams-tests-junit4/src/test/java/com/bakdata/fluent_kafka_streams_tests/junit4/WordCountTest.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2019 bakdata GmbH
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.bakdata.fluent_kafka_streams_tests.junit4;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import com.bakdata.fluent_kafka_streams_tests.junit4.test_applications.WordCount;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.Serdes;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class WordCountTest {
```

Comment:
is there no way to avoid this massive code duplication? something like a base test and the only thing we change in each instance is how we register the `TestTopology`?

along the lines of:

```java
abstract class BaseWordCountTest {
  private TestTopology<> testTopology;  // set this in each subclass
  
  // all test code in here
  @Test
    public void shouldAggregateSameWordStream() { ... }
  ...
}
```

and then 

```java
class Junit5WordCountTest {
  @RegisterExtension
  this.testTopology = new TestTopologyExtension ...
}
```

i have no idea if this is possible but i would strongly prefer it over duplicating all test code 3 times.

Reply:
The main issue is that to properly test junit4 version, no dependencies to junit5 should be in the classpath. @Test is both defined in junit4 and junit5 and although the semantics are the same, they are incompatible.

The only thing that could be shared is the WordCount fixture, but I think the effort is not worth the benefit (if there is any).